src/libraries/LibTokenInfo.sol

```diff
@@ -0,0 +1,464 @@
+  error ErrERC20MintingFailed();
+
+  /// @dev Error indicating that the minting of ERC721 tokens has failed.
+  error ErrERC721MintingFailed();
+
+  /// @dev Error indicating that the transfer of ERC1155 tokens has failed.
+  error ErrERC1155TransferFailed();
+
+  /// @dev Error indicating that the mint of ERC1155 tokens in batch has failed.
+  error ErrERC1155MintBatchFailed();
+
+  /// @dev Error indicating that an unsupported standard is encountered.
+  error ErrUnsupportedStandard();
+
+  /**
+   * @dev Error indicating that the `transfer` has failed.
+   * @param tokenInfo Info of the token including ERC standard, id or quantity.
+   * @param to Receiver of the token value.
+   * @param token Address of the token.
+   */
+  error ErrTokenCouldNotTransfer(TokenInfo tokenInfo, address to, address token);
+
+  /**
+   * @dev Error indicating that the `handleAssetIn` has failed.
+   * @param tokenInfo Info of the token including ERC standard, id or quantity.
+   * @param from Owner of the token value.
+   * @param to Receiver of the token value.
+   * @param token Address of the token.
+   */
+  error ErrTokenCouldNotTransferFrom(TokenInfo tokenInfo, address from, address to, address token);
+
+  /**
+   *
+   *        ROUTER
+   *
+   */
+  function _isModeSingle(Mode mode) private pure returns (bool) {
+    return mode == Mode.Single;
+  }
+
+  function _isModeBatch(Mode mode) private pure returns (bool) {
+    return mode == Mode.Batch;
+  }
+
+  /**
+   *
+   *        HASH
+   *
+   */
+
+  // keccak256("TokenInfo(uint8 erc,uint256 id,uint256 quantity)");
+  bytes32 public constant INFO_TYPE_HASH_SINGLE = 0x1e2b74b2a792d5c0f0b6e59b037fa9d43d84fbb759337f0112fcc15ca414fc8d;
+
+  // keccak256("TokenInfo(uint8 erc,uint256[] ids,uint256[] quantities)");
+  bytes32 public constant INFO_TYPE_HASH_BATCH = 0xe0d9a8bb18cfc29aa6e46b1293275ca79aeaaf28ac63b66dcb6ebce2f127f5a0;
+
+  /**
+   * @dev Returns token info struct hash.
+   */
+  function hash(TokenInfo memory self) internal pure returns (bytes32 digest) {
+    if (_isModeSingle(self.mode)) return _hashSingle(self);
+    if (_isModeBatch(self.mode)) return _hashBatch(self);
+    revert ErrUnsupportedStandard();
+  }
+
+  function _hashSingle(TokenInfo memory self) internal pure returns (bytes32 digest) {
+    // return keccak256(abi.encode(INFO_TYPE_HASH_SINGLE, self.erc, self.id, self.quantity));
+    assembly {
+      let ptr := mload(0x40)
+      mstore(ptr, INFO_TYPE_HASH_SINGLE)
+      mstore(add(ptr, 0x20), mload(self)) // info.erc
+      mstore(add(ptr, 0x40), mload(add(self, 0x20))) // info.id
+      mstore(add(ptr, 0x60), mload(add(self, 0x40))) // info.quantity
+      digest := keccak256(ptr, 0x80)
+    }
+  }
+
+  function _hashBatch(TokenInfo memory self) internal pure returns (bytes32 digest) {
+    bytes32 idsHash = keccak256(abi.encode(self.ids));
+    bytes32 qtysHash = keccak256(abi.encode(self.quantities));
```

Comment:
```suggestion
    bytes32 idsHash = keccak256(abi.encodePacked(self.ids));
    bytes32 qtysHash = keccak256(abi.encodePacked(self.quantities));
```

Reply:
this can be further optimized by using array lib

Reply:
what if there is hash collision?

Reply:
need double check with FE